File: backend/api/routes/face.py
"""Face enroll and recognize API."""
import base64
import logging
import os
import numpy as np
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse

logger = logging.getLogger("face")

from config import SIMILARITY_THRESHOLD, MIN_MARGIN, DATA_DIR, MIN_ENROLLMENTS_FOR_ATTENDANCE
from services.face_service import (
    get_embedding_from_base64,
    get_embedding_from_base64_debug,
    embedding_similarity,
    embedding_to_similarity,
    model_order_for_dim,
)
from repositories.embedding_store import (
    add_embedding,
    get_embeddings,
    get_count,
    remove_all,
    remove_by_index,
    get_all_for_class,
)
from schemas.face import (
    EnrollRequest,
    EnrollResponse,
    RecognizeRequest,
    RecognizeResponse,
    CountResponse,
    EnrolledStudentsResponse,
    DebugImageRequest,
)

router = APIRouter()


def _check_duplicate(user_id: str, class_id: str, embedding: list[float], exclude_student_id: str) -> tuple[str, float] | None:
    """If embedding matches another student, return (student_id, similarity).
    ตรวจสอบที่ 95% (0.95) เพื่อแจ้งเตือนผู้ใช้"""
    dim = len(embedding)
    # ใช้ threshold 95% สำหรับการแจ้งเตือน (สูงกว่า threshold ปกติ)
    # เพื่อให้แจ้งเตือนเฉพาะกรณีที่คล้ายกันมากจริงๆ
    if dim == 128:
        threshold = 0.95  # สำหรับ 128-dim models
    elif dim == 4096:
        threshold = 0.95  # สำหรับ 4096-dim models
    else:
        threshold = 0.95  # สำหรับ Facenet512 (512-dim) และอื่นๆ
    candidates = get_all_for_class(user_id, class_id)
    for student_id, embs in candidates:
        if student_id == exclude_student_id:
            continue
        for emb in embs:
            sim = embedding_similarity(embedding, emb)
            # แปลง similarity เป็น percentage (0-1) แล้วเปรียบเทียบกับ 0.95
            similarity_score = embedding_to_similarity(sim, dim)
            if similarity_score >= threshold:
                return (student_id, similarity_score)
    return None


def _get_existing_dim_for_student(user_id: str, class_id: str, student_id: str) -> int | None:
    """Return embedding dimension already stored for a student (first found)."""
    records = get_embeddings(user_id, class_id, student_id)
    for r in records:
        emb = r.get("embedding")
        if emb:
            return len(emb)
    return None


def _get_dims_for_class(user_id: str, class_id: str) -> set[int]:
    """Return all embedding dimensions stored for a class."""
    dims: set[int] = set()
    candidates = get_all_for_class(user_id, class_id)
    for _, embs in candidates:
        for emb in embs:
            if emb:
                dims.add(len(emb))
    return dims


@router.get("/ping")
def ping():
    """ทดสอบว่า Backend เชื่อมต่อได้"""
    return {"ok": True, "message": "Backend เชื่อมต่อได้", "endpoint": "face-api"}


@router.post("/debug-image")
def debug_image(req: DebugImageRequest):
    """ทดสอบว่า backend สามารถรับรูปภาพได้ (ไม่บันทึกลง disk เพื่อความปลอดภัย)"""
    try:
        s = req.image_base64.strip()
        if "," in s and s.startswith("data:"):
            s = s.split(",", 1)[1]
        raw = base64.b64decode(s, validate=False)
        # ไม่บันทึกรูปภาพลง disk เพื่อความปลอดภัยและความเป็นส่วนตัวของนักเรียน
        # แค่ตรวจสอบว่า decode ได้หรือไม่
        return {"ok": True, "size_bytes": len(raw), "message": "รูปภาพรับได้ (ไม่บันทึกเพื่อความปลอดภัย)"}
    except Exception as e:
        return {"ok": False, "error": str(e)}


@router.post("/debug-extract")
def debug_extract(req: DebugImageRequest):
    """ทดสอบ extract embedding และ return error details เพื่อ debug"""
    return get_embedding_from_base64_debug(req.image_base64 or "")


@router.post("/enroll", response_model=EnrollResponse)
def enroll(req: EnrollRequest):
    try:
        print(
            f"\n>>> [ENROLL] request received - user={req.user_id} class={req.class_id} "
            f"student={req.student_id} image_len={len(req.image_base64 or '')}"
        )
        logger.info("POST /enroll received — user=%s class=%s student=%s image_len=%d", req.user_id, req.class_id, req.student_id, len(req.image_base64 or ""))
        existing_dim = _get_existing_dim_for_student(req.user_id, req.class_id, req.student_id)
        preferred_models = model_order_for_dim(existing_dim) if existing_dim else None
        result = get_embedding_from_base64(req.image_base64, preferred_models=preferred_models)
        if not result:
            debug = get_embedding_from_base64_debug(req.image_base64)
            # ไม่บันทึกรูปภาพลง disk เพื่อความปลอดภัยและความเป็นส่วนตัวของนักเรียน
            # มี debug info ใน response แล้ว (image_dims, errors)
            return JSONResponse(status_code=400, content={"detail": "ไม่พบใบหน้าในภาพ", "debug": debug})
        emb, conf = result
        if existing_dim and len(emb) != existing_dim:
            remove_all(req.user_id, req.class_id, req.student_id)
            logger.info("dim ไม่ตรง (expected=%s got=%s): ล้าง embedding เก่าอัตโนมัติ user=%s class=%s student=%s", existing_dim, len(emb), req.user_id, req.class_id, req.student_id)
            print(f">>> [ENROLL] โมเดลไม่ตรง (expected dim={existing_dim}, got dim={len(emb)}) — ล้างข้อมูลเก่าอัตโนมัติแล้ว user={req.user_id} class={req.class_id} student={req.student_id}")
        dup = _check_duplicate(req.user_id, req.class_id, emb, req.student_id)
        if dup and not req.allow_duplicate:
            other_id, sim = dup
            return JSONResponse(
                status_code=409,
                content={
                    "detail": "ใบหน้านี้ใกล้เคียงกับนักเรียนคนอื่น ต้องการยืนยันการลงทะเบียนหรือไม่",
                    "duplicate": {"student_id": other_id, "similarity": sim},
                },
            )
        count = get_count(req.user_id, req.class_id, req.student_id)
        if count >= 5:
            raise HTTPException(status_code=400, detail="มีข้อมูลใบหน้าครบ 5 รายการแล้ว")
        add_embedding(req.user_id, req.class_id, req.student_id, emb, conf)
        return EnrollResponse(success=True, count=count + 1, message="ลงทะเบียนสำเร็จ")
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("ENROLL failed: %s", e)
        print(f">>> [ENROLL] ERROR: {type(e).__name__}: {e}")
        raise HTTPException(status_code=500, detail=f"ลงทะเบียนล้มเหลว: {type(e).__name__}: {str(e)}")


@router.post("/recognize", response_model=RecognizeResponse)
def recognize(req: RecognizeRequest):
    from repositories.embedding_store import get_normalized_embeddings_for_class
    
    # ดึง embedding จากรูปสแกนแบบเดียวกับตอนลงทะเบียน (ไม่บังคับโมเดล = ใช้ mediapipe + Facenet512)
    # เพื่อให้จับคู่ได้กับข้อมูลที่ลงทะเบียนใหม่ (512)
    result = get_embedding_from_base64(req.image_base64, preferred_models=None)
    if not result:
        return RecognizeResponse(student_id=None, student_name=None, similarity=0, matched=False)
    query_emb, _ = result
    query_dim = len(query_emb)
    threshold = 0.4 if query_dim == 128 else SIMILARITY_THRESHOLD

    # Get pre-normalized embeddings cache (much faster!)
    # Only match against students with at least MIN_ENROLLMENTS_FOR_ATTENDANCE images
    normalized_by_dim = get_normalized_embeddings_for_class(
        req.user_id, req.class_id, min_embeddings=MIN_ENROLLMENTS_FOR_ATTENDANCE
    )
    if query_dim not in normalized_by_dim or not normalized_by_dim[query_dim]:
        return RecognizeResponse(student_id=None, student_name=None, similarity=0, matched=False)

    # Convert query to numpy array and normalize once
    query_arr = np.array(query_emb, dtype=np.float32)
    query_norm = np.linalg.norm(query_arr)
    if query_norm == 0:
        return RecognizeResponse(student_id=None, student_name=None, similarity=0, matched=False)
    query_normalized = query_arr / query_norm

    best_student_id = None
    best_similarity = 0.0
    second_best_similarity = 0.0

    # Ultra-fast vectorized computation using pre-normalized embeddings
    for student_id, normalized_matrix, _ in normalized_by_dim[query_dim]:
        # normalized_matrix is already (n_embeddings, dim) and normalized
        # Compute all similarities at once (vectorized)
        similarities = np.dot(normalized_matrix, query_normalized)
        
        # Find best match for this student
        max_sim = float(np.max(similarities))
        
        if max_sim > best_similarity:
            second_best_similarity = best_similarity
            best_similarity = max_sim
            best_student_id = student_id
        elif max_sim > second_best_similarity:
            second_best_similarity = max_sim
        
        # Early exit if we found a very good match (ultra-fast path)
        # เพิ่ม threshold สำหรับ early exit เพื่อความแม่นยำสูงขึ้น
        from config import HIGH_CONFIDENCE_THRESHOLD
        if best_similarity > HIGH_CONFIDENCE_THRESHOLD:  # Very high confidence (0.75)
            break

    if best_student_id is None:
        return RecognizeResponse(student_id=None, student_name=None, similarity=0, matched=False)

    display_sim = embedding_to_similarity(best_similarity, query_dim)
    
    # เพิ่มความแม่นยำ: ตรวจสอบ threshold ที่เข้มงวดขึ้น
    if best_similarity < threshold:
        return RecognizeResponse(
            student_id=None, student_name=None, similarity=display_sim, matched=False
        )
    
    # เพิ่มความแม่นยำ: ตรวจสอบ margin ที่เข้มงวดขึ้นเพื่อแยกบุคคลได้ชัดเจน
    if second_best_similarity > 0 and (best_similarity - second_best_similarity) < MIN_MARGIN:
        return RecognizeResponse(
            student_id=None, student_name=None, similarity=display_sim, matched=False
        )
    
    # เพิ่มการตรวจสอบ: ถ้ามี similarity ใกล้เคียงกันมากเกินไป (อาจเป็นคนเดียวกันหรือคล้ายกันมาก)
    # ให้ตรวจสอบเพิ่มเติมด้วยการเปรียบเทียบกับ embeddings ทั้งหมดของนักเรียนคนนั้น
    # เข้มงวดขึ้น: ป้องกันการจดจำผิดคนเมื่อใบหน้าคล้ายกัน
    if second_best_similarity > 0:
        margin = best_similarity - second_best_similarity
        # ถ้า margin น้อยกว่า 0.20 แสดงว่าอาจมีความคล้ายคลึงกันสูง ต้องแน่ใจว่าเป็นคนเดียวกัน
        if margin < 0.20:
            # ตรวจสอบว่า best_similarity สูงพอหรือไม่
            # ถ้า margin น้อย ต้องมีความมั่นใจสูงมาก
            required_confidence = 0.75 if margin < 0.15 else 0.72
            if best_similarity < required_confidence:
                return RecognizeResponse(
                    student_id=None, student_name=None, similarity=display_sim, matched=False
                )
    
    return RecognizeResponse(
        student_id=best_student_id,
        student_name=None,
        similarity=display_sim,
        matched=True,
    )


@router.get("/count", response_model=CountResponse)
def get_face_count(user_id: str, class_id: str, student_id: str):
    return CountResponse(count=get_count(user_id, class_id, student_id))


@router.get("/enrolled", response_model=EnrolledStudentsResponse)
def get_enrolled_students(user_id: str, class_id: str):
    """Return student IDs that have at least MIN_ENROLLMENTS_FOR_ATTENDANCE face images (can check attendance)."""
    candidates = get_all_for_class(user_id, class_id)
    ids = [s[0] for s in candidates if len(s[1]) >= MIN_ENROLLMENTS_FOR_ATTENDANCE]
    return EnrolledStudentsResponse(student_ids=ids)


@router.delete("/enroll")
def delete_enrollment(user_id: str, class_id: str, student_id: str, index: int | None = None):
    if index is not None:
        remove_by_index(user_id, class_id, student_id, index)
    else:
        remove_all(user_id, class_id, student_id)
    return {"success": True}


@router.get("/list")
def list_enrollments(user_id: str, class_id: str, student_id: str):
    records = get_embeddings(user_id, class_id, student_id)
    return {
        "count": len(records),
        "records": [{"enrolledAt": r.get("enrolledAt", ""), "confidence": r.get("confidence", 0)} for r in records],
    }

File: app/src/api/face.ts
/**
 * Face API Backend - Calls Python backend (RetinaFace + InsightFace)
 */

const API_BASE = import.meta.env.VITE_API_URL || 'http://localhost:8000';

export async function checkBackendHealth(): Promise<boolean> {
  try {
    const res = await fetch(`${API_BASE}/api/health`);
    return res.ok;
  } catch {
    return false;
  }
}

/** ทดสอบการเชื่อมต่อ Frontend ↔ Backend แบบละเอียด */
export async function testConnection(imageBase64?: string): Promise<{
  ok: boolean;
  health: boolean;
  saveImage: boolean | null;
  error?: string;
}> {
  const result: { ok: boolean; health: boolean; saveImage: boolean | null; error?: string } = { ok: false, health: false, saveImage: null };
  try {
    const healthRes = await fetch(`${API_BASE}/api/health`);
    result.health = healthRes.ok;
    if (!result.health) {
      result.error = 'Backend ไม่ตอบ (health 404)';
      return result;
    }
    if (imageBase64) {
      try {
        const saveRes = await fetch(`${API_BASE}/api/face/debug-image`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ image_base64: imageBase64 }),
        });
        const saveData = await saveRes.json().catch(() => ({}));
        result.saveImage = saveRes.ok && !saveData.error;
        if (!result.saveImage) result.error = (result.error || '') + ` | บันทึกรูป: ${saveData.error || 'ไม่สำเร็จ'}`;
      } catch {
        result.saveImage = false;
        result.error = (result.error || '') + ' | บันทึกรูป: network error';
      }
    }
    result.ok = result.health && (result.saveImage !== false);
  } catch (e) {
    result.error = e instanceof Error ? e.message : String(e);
  }
  return result;
}

export async function debugExtract(imageBase64: string): Promise<{ ok: boolean; errors?: string[]; image_size?: number; image_dims?: string }> {
  const res = await fetch(`${API_BASE}/api/face/debug-extract`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ image_base64: imageBase64 }),
  });
  return res.json();
}

export type EnrollResult =
  | { success: true; count: number }
  | { success: false; reason: 'duplicate'; duplicate: { student_id: string; similarity: number } };

export async function enrollFace(
  classId: string,
  studentId: string,
  imageBase64: string,
  options?: { allowDuplicate?: boolean; forceNewModel?: boolean }
): Promise<EnrollResult> {
  console.log('[enrollFace] ส่ง request ไป', `${API_BASE}/api/face/enroll`, 'image_len=', imageBase64?.length);
  let res: Response;
  try {
    res = await fetch(`${API_BASE}/api/face/enroll`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        class_id: classId,
        student_id: studentId,
        image_base64: imageBase64,
        allow_duplicate: options?.allowDuplicate ?? false,
        force_new_model: options?.forceNewModel ?? false,
      }),
    });
  } catch (err) {
    console.error('[enrollFace] Fetch failed (backend ไม่ตอบ?):', err);
    throw new Error('เชื่อมต่อ Backend ไม่ได้ — กรุณารัน Backend ก่อน');
  }
  const data = await res.json().catch(() => ({}));
  if (res.status === 409 && data?.duplicate) {
    return { success: false, reason: 'duplicate', duplicate: data.duplicate };
  }
  if (!res.ok) {
    if (data.debug) {
      console.log('[enrollFace] Backend debug:', data.debug);
      const errors = data.debug.errors || [];
      if (errors.length > 0) {
        console.error('[enrollFace] Extraction errors:', errors);
      }
    }
    const msg = Array.isArray(data.detail) ? data.detail[0]?.msg ?? data.detail[0] : data.detail;
    const errorMsg = typeof msg === 'string' ? msg : 'ลงทะเบียนไม่สำเร็จ';
    const debugInfo = data.debug ? ` (image: ${data.debug.image_dims || 'unknown'}, errors: ${(data.debug.errors || []).join('; ')})` : '';
    throw new Error(errorMsg + debugInfo);
  }
  return { success: true, count: data.count };
}

export interface RecognizeResult {
  student_id: string | null;
  student_name: string | null;
  similarity: number;
  matched: boolean;
}

export async function recognizeFace(
  classId: string,
  imageBase64: string,
  signal?: AbortSignal
): Promise<RecognizeResult> {
  const res = await fetch(`${API_BASE}/api/face/recognize`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      class_id: classId,
      image_base64: imageBase64,
    }),
    signal, // Support request cancellation
  });
  const data = await res.json().catch(() => ({}));
  if (!res.ok) {
    const msg = Array.isArray(data.detail) ? data.detail[0]?.msg ?? data.detail[0] : data.detail;
    throw new Error(typeof msg === 'string' ? msg : 'การยืนยันตัวตนล้มเหลว');
  }
  return {
    student_id: data.student_id ?? null,
    student_name: data.student_name ?? null,
    similarity: data.similarity ?? 0,
    matched: data.matched ?? false,
  };
}

export async function getFaceCount(
  classId: string,
  studentId: string
): Promise<number> {
  try {
    const res = await fetch(
      `${API_BASE}/api/face/count?class_id=${encodeURIComponent(classId)}&student_id=${encodeURIComponent(studentId)}`
    );
    const data = await res.json().catch(() => ({}));
    return data.count ?? 0;
  } catch {
    return 0;
  }
}

export async function getEnrolledStudentIds(classId: string): Promise<string[]> {
  try {
    const res = await fetch(
      `${API_BASE}/api/face/enrolled?class_id=${encodeURIComponent(classId)}`
    );
    const data = await res.json().catch(() => ({}));
    return data.student_ids ?? [];
  } catch {
    return [];
  }
}

export async function removeFaceEnrollment(
  classId: string,
  studentId: string,
  index?: number
): Promise<void> {
  let url = `${API_BASE}/api/face/enroll?class_id=${encodeURIComponent(classId)}&student_id=${encodeURIComponent(studentId)}`;
  if (index != null) url += `&index=${index}`;
  const res = await fetch(url, { method: 'DELETE' });
  if (!res.ok) {
    const text = await res.text();
    let msg = 'ลบไม่สำเร็จ';
    try {
      const data = JSON.parse(text);
      if (typeof data.detail === 'string') msg = data.detail;
    } catch {
      if (text) msg = `${msg} (${res.status})`;
    }
    throw new Error(msg);
  }
}

export async function getFaceEnrollmentRecords(
  classId: string,
  studentId: string
): Promise<{ enrolledAt: string; confidence: number }[]> {
  try {
    const res = await fetch(
      `${API_BASE}/api/face/list?class_id=${encodeURIComponent(classId)}&student_id=${encodeURIComponent(studentId)}`
    );
    const data = await res.json().catch(() => ({}));
    return data.records ?? [];
  } catch {
    return [];
  }
}
